core/src/main/scala/ox/channels/SourceOps.scala

```
@@ -1008,4 +1009,114 @@ trait SourceOps[+T] { outer: Source[T] =>
+    */
+  def alsoTo[U >: T](other: Sink[U])(using Ox, StageCapacity): Source[U] =
+    val c2 = StageCapacity.newChannel[U]
+    fork {
+      repeatWhile {
+        receiveOrClosed() match
+          case ChannelClosed.Done =>
+            c2.doneOrClosed()
+            other.doneOrClosed()
+            false
+          case ChannelClosed.Error(r) =>
+            c2.errorOrClosed(r)
+            other.errorOrClosed(r)
+            false
+          case t: T @unchecked =>
+            if c2.sendOrClosed(t).isValue then
+              if other.sendOrClosed(t).isValue then true
+              else
+                c2.doneOrClosed().discard
+                false
+            else
+              other.doneOrClosed().discard
+              false
+      }
+    }
+    c2
+
+  /** Attaches the given Sink to this Source, meaning that elements that pass through will also be sent to the Sink. The elements sent to
+    * `other` Sink are enqueued, which means that the returned channel does not block when the `other` Sink is blocked. If the queue is
+    * full, the elements are still sent to returned channel, but not to the `other` Sink, meaning that some elements may be dropped.
+    *
+    * If this source is failed then failure is passed to the returned and channel and the `other` Sink.
+    *
+    * @param other
+    *   The Sink to which elements from this source will be sent.
+    * @param queueSize
+    *   Specifies the size of the queue that buffers elements sent to the `other` Sink.
+    * @return
+    *   A source that emits input elements.
+    * @example
+    *   {{{
+    *   import ox.*
+    *   import ox.channels.Source
+    *
+    *   supervised {
+    *     val c = Channel.withCapacity[Int](10)
+    *     Source.fromValues(1, 2, 3).wireTap(c).toList // List(1, 2, 3)
+    *     c.toList // List(1, 2, 3) but not guaranteed
+    *   }
+    *   }}}
+    */
+  def wireTap[U >: T](other: Sink[U], queueSize: Long = 3)(using Ox, StageCapacity): Source[U] =
```

so ... wireTap is like `alsoTo`, but with a buffer, and dropping elements if the buffer is full?

wouldn't a "proper" solution in such case be to just use an `other` sink with a larger capacity? Because it seems we are doing double-buffering here.

The option to drop elements if `other` is full might be useful of course, as an optional parameter to `alsoTo`?

---

Another difference is how the completion of `other` is handled. With `wireTap` we shouldn't care (as we're discussing here https://github.com/softwaremill/ox/pull/166#discussion_r1658456448). With `alsoTo` completion of `other` should complete `c2`.

---

ah good point :) still, these two look so similar, but have such different names ... ;)

---

> wouldn't a "proper" solution in such case be to just use an other sink with a larger capacity? 
I don't think so. Suppose we have an web application with attached external metric collecting system that goes down sometimes or is very slow, but we still want to process the requests as fast as possible. `wireTap` seems a good choice for this scenario, because it simply drops these metric updates instead of buffering them.

The proposed implementation introduces some kind of logical queue, but we could probably extend jox api with something like `c.sendOrIgnoreIfFull(t)`

---

`c.sendOrIgnoreIfFull == select(c.sendClause(t), Default(NotSent))`

---

but we could also add this as a `trySend` method or sth similar with a name not suggesting exception-handling

---

this `select` is a game changer :)
one more thing regarding `wireTap`: should we propagate the error from `other` to `c2`?

---

cc @kciesielski

---

I assumed that we should, but Akka's wiretap documentation uses the expression 
> elements that pass through will also be sent to the wire-tap Sink, without the latter affecting the mainline flow
and now I'm not sure :) I think it's unsafe to let Sink fail silently. Could you test what happens in your test if you fail it on akka-streams after it processes a few elements?

---

> Could you test what happens in your test if you fail it on akka-streams after it processes a few elements?

after replacing `completeStage()` with `failStage(new RuntimeException("Boom!"))` in `wireTap` sink logic, it works the same way - elements pass through until source completes

---

Akka's model is a bit different because they have "cold" streams - they run the entire pipeline at once (when it's materialised). Here we've got "live" channels that we are operating on, which could be consumed by anybody anywhere. So if "tap" is supposed to be "best-effort", then if `other` is closed, I think we should just ignore it?

---

Hmm I understand this makes the attached secondary sink really not-affecting the mainline at all. Maybe that's what we need then, a true "silent listener". The `other` sink errors would be lost to the user, unless they add some handling themselves, right? Do akka-streams log that "Boom"?

---

They would not be lost, they would surface whenever you try to receive from the secondary sink